### DISCUSSION TITLE
Cache in spike prime

### DISCUSSION THREAD

**Discussion by techcatgato** [pybricks/support#1749] [Q&A]
Does spike prime have any cache that might be able to be utilized by my PyBricks programs?

**Comment by laurensvalk** [comment]
You can make arrays or matrices to keep data in memory.

If you need a small amount of persistent data, you can use the hub.system.storage method.

**Reply by techcatgato** [reply]
If I have a mathematical function that will be used a lot (10s of times per second probably) and needs to be executed fast, will it be benifical to use the method for it? My main concern here is speed which I want to be as fast as possible

**Reply by laurensvalk** [reply]
10s of times per second should be fine without any special techniques.

The technique I mentioned above is for storing data like sensor values or settings, not functions.

**Reply by techcatgato** [reply]
> 10s of times per second should be fine without any special techniques.
> 
> The technique I mentioned above is for storing data like sensor values or settings, not functions.

It can be simplified as a string though. My question is will it make any difference in speed? even 1% is good I want it as fast as possible

**Reply by laurensvalk** [reply]
It's hard to say without knowing what you want to do. 

If it has anything to do with motors (like line following), then math calculation time is usually negligible compared to the time scale to move a motor.

Storing data isn't going to speed anything up. I just thought you originally asked about caching some data for future use, but it sounds like you are doing something different.

**Reply by techcatgato** [reply]
> It's hard to say without knowing what you want to do.
> 
> If it has anything to do with motors (like line following), then math calculation time is usually negligible compared to the time scale to move a motor.
> 
> Storing data isn't going to speed anything up. I just thought you originally asked about caching some data for future use, but it sounds like you are doing something different.

Ok, let me explain. I have a mathematical function (input x output y) which I will need to execute a lot while moving, and i want it to be computed as fast as possible to always get the most fitting value of y according to x which changes linearly (although the relationship isnt linear). To do this I will do something like y = eval("function in string form") and then use y for my program many times a second. To show you the complexity of the functions I will use, Here is an example for one (in python readable format and order of operations):
`-2.7620461494187668e-18*x**5 + 8.6770520698389e-17*x**4 + -1.0937235414935658e-15*x**3 + 7.418489234551116e-15*x**2 + 1.9999999999999762*x + 5.000000000000035`

**Reply by dlech** [reply]
`eval` is going to be much slower than just writing a function because it has to compile the entire expression each time it is called. It would be better to just write a function instead.

```python
def my_func(x):
    return -2.7620461494187668e-18*x**5 + 8.6770520698389e-17*x**4 + -1.0937235414935658e-15*x**3 + 7.418489234551116e-15*x**2 + 1.9999999999999762*x + 5.000000000000035
```

But the best way to figure out how long a function is actually taking is to measure it. To get more accurate measurements, you can call the function many 1000s of times.

```python
from pybricks.tools import StopWatch

COUNT = 10000

def my_func(x):
    return (
        -2.7620461494187668e-18 * x**5
        + 8.6770520698389e-17 * x**4
        + -1.0937235414935658e-15 * x**3
        + 7.418489234551116e-15 * x**2
        + 1.9999999999999762 * x
        + 5.000000000000035
    )

timer = StopWatch()

for n in range(COUNT):
    y = my_func(n / COUNT)

time = timer.time()

print(f"my_func took {time / COUNT} ms")
```

You can use this to test different functions to find which one is the fastest.

**Reply by techcatgato** [reply]
> `eval` is going to be much slower than just writing a function because it has to compile the entire expression each time it is called. It would be better to just write a function instead.
> 
> ```python
> def my_func(x):
>     return -2.7620461494187668e-18*x**5 + 8.6770520698389e-17*x**4 + -1.0937235414935658e-15*x**3 + 7.418489234551116e-15*x**2 + 1.9999999999999762*x + 5.000000000000035
> ```
> 
> But the best way to figure out how long a function is actually taking is to measure it. To get more accurate measurements, you can call the function many 1000s of times.
> 
> ```python
> from pybricks.tools import StopWatch
> 
> COUNT = 10000
> 
> def my_func(x):
>     return (
>         -2.7620461494187668e-18 * x**5
>         + 8.6770520698389e-17 * x**4
>         + -1.0937235414935658e-15 * x**3
>         + 7.418489234551116e-15 * x**2
>         + 1.9999999999999762 * x
>         + 5.000000000000035
>     )
> 
> timer = StopWatch()
> 
> for n in range(COUNT):
>     y = my_func(n / COUNT)
> 
> time = timer.time()
> 
> print(f"my_func took {time / COUNT} ms")
> ```
> 
> You can use this to test different functions to find which one is the fastest.

Okay thank you, found out I dont need to use either eval() or {} because the function accepts the mathematical function as a valid parameter. Are there any further ways I can optimize the code to run faster?

**Reply by dlech** [reply]
I would not expect that using `{}` would be a major slowdown. But the only way to know for sure is to try it and measure it. :smile:

**Reply by techcatgato** [reply]
> I would not expect that using `{}` would be a major slowdown. But the only way to know for sure is to try it and measure it. 😄

Okay, thank you. Is there any further optimization I could apply to this code? It will involve motors so optimizatino for that would help too.

**Reply by dlech** [reply]
When I measured it, I got: `my_func took 0.1782 ms`, so it seems already quite fast. We could probably knock off a few more microseconds by avoiding the exponents, but I don't think that would make much of difference in your program if you are only calling this 10 or 100 times per second.

**Reply by techcatgato** [reply]
> When I measured it, I got: `my_func took 0.1782 ms`, so it seems already quite fast. We could probably knock off a few more microseconds by avoiding the exponents, but I don't think that would make much of difference in your program if you are only calling this 10 or 100 times per second.

When I send the output to a motor Is it still going to be fast? Dont have a motor on me right now so I cant check

**Reply by techcatgato** [reply]
> When I measured it, I got: `my_func took 0.1782 ms`, so it seems already quite fast. We could probably knock off a few more microseconds by avoiding the exponents, but I don't think that would make much of difference in your program if you are only calling this 10 or 100 times per second.

Did you run yours on your PC or on the SPIKE PRIME?

**Reply by dlech** [reply]
On the SPIKE Prime.

**Reply by laurensvalk** [reply]
> > When I measured it, I got: `my_func took 0.1782 ms`, so it seems already quite fast. We could probably knock off a few more microseconds by avoiding the exponents, but I don't think that would make much of difference in your program if you are only calling this 10 or 100 times per second.
> 
> When I send the output to a motor Is it still going to be fast? Dont have a motor on me right now so I cant check

Sending it to the motor is almost instant. Especially if you control acceleration yourself using the dc() method.

But it is only going to respond as fast as the mechanical components in the motor can move. Usually with this type of system, any computation under 10 ms isn't going to slow things down all that much.

For reference, a self balancing robot needs to respond pretty quickly to stay balanced but it still works great with a 20ms or even a 50 ms loop time.

More formally speaking, a LEGO motor (or any similarly sized electro mechanical system) acts as a low pass filter. Even if you throw very fast changing data at it, it only cares about the average value over time.
